driver/src/orderbook/streamed/state.rs

```diff
@@ -155,6 +165,91 @@ impl State {
         // order.
         Ok(())
     }
+
+    fn apply_trade(&mut self, event: &Trade, block_batch_id: BatchId) -> Result<()> {
+        self.apply_trade_internal(
+            event.owner,
+            event.order_id,
+            |order| order.trade(event.executed_sell_amount, block_batch_id),
+            |sell_balance| sell_balance.sell(event.executed_sell_amount, block_batch_id),
+            |buy_balance| buy_balance.buy(event.executed_buy_amount, block_batch_id),
+        )
+    }
+
+    fn apply_trade_reversion(
+        &mut self,
+        event: &TradeReversion,
+        block_batch_id: BatchId,
+    ) -> Result<()> {
+        self.apply_trade_internal(
+            event.owner,
+            event.order_id,
+            |order| order.revert_trade(event.executed_sell_amount, block_batch_id),
+            |sell_balance| sell_balance.revert_sell(event.executed_sell_amount, block_batch_id),
+            |buy_balance| buy_balance.revert_buy(event.executed_buy_amount, block_batch_id),
+        )
+    }
+
+    fn apply_trade_internal(
+        &mut self,
+        user_id: UserId,
+        order_id: OrderId,
+        order_fn: impl FnOnce(&mut Order) -> Result<()>,
+        sell_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
+        buy_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
+    ) -> Result<()> {
+        let order = self
+            .orders
+            .get_mut(&(user_id, order_id))
+            .ok_or_else(|| anyhow!("unknown order"))?;
+        order_fn(order)?;
+
+        let sell_token = self
+            .tokens
+            .get_address_by_id(order.sell_token)
+            .ok_or_else(|| anyhow!("unknown sell token"))?;
+        let sell_balance = self.balances.entry((user_id, sell_token)).or_default();
+        sell_balance_fn(sell_balance)?;
+
+        let buy_token = self
+            .tokens
+            .get_address_by_id(order.buy_token)
+            .ok_or_else(|| anyhow!("unknown buy token"))?;
+        let buy_balance = self.balances.entry((user_id, buy_token)).or_default();
+        buy_balance_fn(buy_balance)
+    }
+
+    fn apply_solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<()> {
+        let fee_token = self
+            .tokens
+            .get_address_by_id(0)
+            .ok_or_else(|| anyhow!("solution without fee token"))?;
+        self.revert_last_solution(fee_token, block_batch_id);
+        self.last_solution.batch_id = block_batch_id;
+        self.last_solution.user_id = event.submitter;
+        self.last_solution.burnt_fees = event.burnt_fees;
+        self.balances
+            .entry((event.submitter, fee_token))
+            .or_default()
+            .solution_submission(event.burnt_fees, block_batch_id)
+    }
```

Comment:
I've seen this already in a previously merged PR.

Reply:
Not in a merged PR. We cancelled that one because it was too big iirc.